File: simple-stack-example-mvp/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/addoredittask/AddOrEditTaskKey.java
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.addoredittask;

import android.os.Parcelable;
import android.view.View;

import com.google.auto.value.AutoValue;
import com.squareup.coordinators.Coordinator;
import com.squareup.coordinators.Coordinators;
import com.zhuinden.simplestackdemoexamplemvp.R;
import com.zhuinden.simplestackdemoexamplemvp.application.Key;
import com.zhuinden.simplestackdemoexamplemvp.application.injection.SingletonComponent;
import com.zhuinden.simplestackdemoexamplemvp.presentation.paths.tasks.TasksKey;

/**
 * Created by Zhuinden on 2017.01.25..
 */

@AutoValue
public abstract class AddOrEditTaskKey
        implements Key {
    public abstract TasksKey parent();

    @Override
    public Coordinator newCoordinator(SingletonComponent singletonComponent) {
        return singletonComponent.addOrEditTaskCoordinator();
    }

    public static Parcelable create(TasksKey parent) {
        return new AutoValue_AddOrEditTaskKey(R.layout.path_addoredittask, parent);
    }

    @Override
    public int menu() {
        return R.menu.empty_menu;
    }

    @Override
    public boolean isFabVisible() {
        return true;
    }

    @Override
    public int navigationViewId() {
        return 0;
    }

    @Override
    public boolean shouldShowUp() {
        return true;
    }

    @Override
    public View.OnClickListener fabClickListener(View view) {
        return v -> {
            AddOrEditTaskCoordinator addOrEditTaskCoordinator = Coordinators.getCoordinator(view);
            addOrEditTaskCoordinator.fabClicked();
        };
    }

    @Override
    public int fabDrawableIcon() {
        return R.drawable.ic_done;
    }
}

File: samples/legacy-architecture-samples/simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/features/addoredittask/AddOrEditTaskKey.kt
package com.zhuinden.simplestackdemoexamplemvp.features.addoredittask

import android.view.View
import com.zhuinden.simplestack.ServiceBinder
import com.zhuinden.simplestack.navigator.ViewChangeHandler
import com.zhuinden.simplestack.navigator.changehandlers.SegueViewChangeHandler
import com.zhuinden.simplestackdemoexamplemvp.R
import com.zhuinden.simplestackdemoexamplemvp.core.navigation.ViewKey
import com.zhuinden.simplestackdemoexamplemvp.data.repository.TaskRepository
import com.zhuinden.simplestackdemoexamplemvp.util.MessageQueue
import com.zhuinden.simplestackextensions.services.DefaultServiceProvider
import com.zhuinden.simplestackextensions.servicesktx.lookup
import kotlinx.parcelize.Parcelize

sealed class AddOrEditTaskKey(val parent: ViewKey, val taskId: String = "") : ViewKey, DefaultServiceProvider.HasServices {
    override fun getScopeTag() = "AddOrEditTask"

    override fun bindServices(serviceBinder: ServiceBinder) {
        with(serviceBinder) {
            addService(AddOrEditTaskView.CONTROLLER_TAG, AddOrEditTaskPresenter(
                lookup<TaskRepository>(),
                lookup<MessageQueue>(),
                backstack
            ))
        }
    }

    @Parcelize
    data class AddTaskKey(val parentKey: ViewKey) : AddOrEditTaskKey(parentKey)

    @Parcelize
    data class EditTaskKey(val parentKey: ViewKey, val taskID: String) : AddOrEditTaskKey(parentKey, taskID)

    override fun layout(): Int = R.layout.path_addoredittask

    override val isFabVisible: Boolean
        get() = true

    override fun viewChangeHandler(): ViewChangeHandler = SegueViewChangeHandler()

    override fun menu(): Int = R.menu.empty_menu

    override fun navigationViewId(): Int = 0

    override fun shouldShowUp(): Boolean = true

    override fun fabClickListener(view: View): View.OnClickListener =
        View.OnClickListener { v ->
            val addOrEditTaskView = view as AddOrEditTaskView
            addOrEditTaskView.fabClicked()
        }

    override fun fabDrawableIcon(): Int = R.drawable.ic_done
}
